File: vi/codes/cpp/chapter_dynamic_programming/climbing_stairs_dfs.cpp
/**
 * File: climbing_stairs_dfs.cpp
 * Created Time: 2023-06-30
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Tìm kiếm */
int dfs(int i) {
    // Đã biết dp[1] và dp[2], trả về chúng
    if (i == 1 || i == 2)
        return i;
    // dp[i] = dp[i-1] + dp[i-2]
    int count = dfs(i - 1) + dfs(i - 2);
    return count;
}

/* Leo cầu thang: Tìm kiếm */
int climbingStairsDFS(int n) {
    return dfs(n);
}

/* Driver Code */
int main() {
    int n = 9;

    int res = climbingStairsDFS(n);
    cout << "Climbing " << n << " stairs has " << res << " solutions" << endl;

    return 0;
}

File: vi/codes/python/chapter_dynamic_programming/climbing_stairs_dfs.py
"""
File: climbing_stairs_dfs.py
Created Time: 2023-06-30
Author: krahets (krahets@163.com)
"""


def dfs(i: int) -> int:
    """Tìm kiếm"""
    # Đã biết dp[1] và dp[2], trả về chúng
    if i == 1 or i == 2:
        return i
    # dp[i] = dp[i-1] + dp[i-2]
    count = dfs(i - 1) + dfs(i - 2)
    return count


def climbing_stairs_dfs(n: int) -> int:
    """Leo cầu thang: Tìm kiếm"""
    return dfs(n)


"""Chương trình chính"""
if __name__ == "__main__":
    n = 9

    res = climbing_stairs_dfs(n)
    print(f"Climbing {n} stairs has {res} solutions")
